libnavigation-ui/src/main/java/com/mapbox/navigation/ui/internal/building/BuildingLayerSupport.kt

```diff
@@ -0,0 +1,99 @@
+import com.mapbox.mapboxsdk.style.expressions.Expression.literal
+import com.mapbox.mapboxsdk.style.expressions.Expression.lt
+import com.mapbox.mapboxsdk.style.layers.Layer
+import com.mapbox.mapboxsdk.style.layers.PropertyValue
+
+/**
+ * This class handles shared code for the [BuildingExtrusionHighlightLayer]
+ * and [BuildingFootprintHighlightLayer] classes which show
+ * [com.mapbox.mapboxsdk.style.layers.FillExtrusionLayer] 3D buildings
+ * and [com.mapbox.mapboxsdk.style.layers.FillLayer] polygons, respectively.
+ * For now, this class is only compatible with the Mapbox Streets v7 vector tile source
+ * because that's what the default Navigation UI SDK styles use.
+ */
+internal class BuildingLayerSupport {
+
+    /**
+     * Gets the correct [Expression.all] expression to show the building
+     * extrusion associated with the query [LatLng].
+     *
+     * @param queryLatLng the [LatLng] to use in determining the building ID to use in the
+     * expression and which building is closest to the coordinate.
+     * @return an [Expression.all] expression
+     */
+    fun getBuildingFilterExpression(queryLatLng: LatLng, mapboxMap: MapboxMap): Expression {
+        return all(
+                eq(get("extrude"), "true"),
+                eq(get("type"), "building"),
+                eq(get("underground"), "false"),
+                eq(id(), literal(getBuildingId(mapboxMap, queryLatLng))),
+                lt(distance(Point.fromLngLat(queryLatLng.longitude, queryLatLng.latitude)),
+                        literal(QUERY_DISTANCE_MAX_METERS))
+        )
+    }
+
+    /**
+     * Gets the specific ID from the building layer [Feature] that the
+     * queryLatLng is within.
+     *
+     * @param mapboxMap the [MapboxMap] to query
+     * @param queryLatLng the [LatLng] to use in determining the building ID to
+     * use in the expression and which building is closest to the
+     * coordinate.
+     * @return the building ID as an integer
+     */
+    private fun getBuildingId(mapboxMap: MapboxMap, queryLatLng: LatLng): Int {
+        val renderedBuildingFootprintFeatures = mapboxMap.queryRenderedFeatures(
+                mapboxMap.projection.toScreenLocation(queryLatLng), BUILDING_LAYER_ID)
+        if (renderedBuildingFootprintFeatures.isNotEmpty()) {
+            renderedBuildingFootprintFeatures[0].id()?.let {
+                return Integer.valueOf(it)
+            }
+        }
+        return 0
+    }
+
+    /**
+     * Sets a new property on a specified layer
+     */
+    fun updateLayerProperty(propertyValueToSet: PropertyValue<*>?, mapboxMap: MapboxMap, layerId: String) {
+        mapboxMap.getStyle { style ->
+            val mapLayer = style.getLayerAs<Layer>(layerId)
+            mapLayer?.setProperties(propertyValueToSet)
+        }
+    }
+
+    /**
+     * Adds a [Layer] to the [MapboxMap]'s style object.
+     */
+    fun addLayerToStyle(mapboxMap: MapboxMap, layerToAdd: Layer, idOfLayerToCheckAbove: String) {
```

Comment:
This looks like a duplicate of MapUtils.addLayerToMap.

Reply:
👍